### ISSUE TITLE
RFC: Add single-image vision-language model support to A2POTrainer

### ISSUE THREAD

**Issue by DimensionSTP** [huggingface/trl#6833]
### Feature request

`A2POTrainer` is currently text-oriented across model loading, processing, Stage 1 optimal-value estimation, filtering, and Stage 2 policy/reference scoring.

I propose a narrowly scoped first implementation of VLM support:

- one image per prompt through an `image` column;
- conversational prompts;
- Transformers generation only;
- same-architecture policy and reference;
- local Python reward functions;
- full-model training;
- no vLLM, tools, video, multi-image, PEFT continuation, or ZeRO-3 claim.

The implementation would reuse established TRL multimodal patterns:

- `create_model_from_path()` for model architecture inference;
- `AutoProcessor` and `ProcessorMixin` detection;
- `prepare_multimodal_messages()` for associating prompts with source images;
- processor-produced multimodal kwargs for generation and model forwards;
- completion-only log-probability masking;
- explicit failure when truncation would remove required image tokens.

### Motivation

On current main, A2PO:

1. loads string model IDs with `AutoModelForCausalLM`;
2. defaults to `AutoTokenizer`;
3. sends rendered text only to the Stage 1 processor and reference generation;
4. sends rendered text only to Stage 2 policy generation;
5. forwards policy and reference models with only `input_ids` and `attention_mask`;
6. keys optimal values and all-incorrect filtering by rendered prompt text.

The last point is an A2PO-specific correctness problem, not only missing multimodal plumbing. These two samples are distinct:

```python
{"prompt": "What is shown?", "image": cat_image}
{"prompt": "What is shown?", "image": invoice_image}
```

They have the same rendered prompt text. The current cache therefore stores the second `V*` under the same key and overwrites the first. `filter_all_incorrect` can likewise let one image's reward result affect the other row.

A CPU/unit reproduction against main at `06c89977492db9102c41a920b499426de6b1c776` independently confirmed five gaps:

- forced causal-LM loader;
- forced tokenizer loader;
- image loss in Stage 1;
- image loss in Stage 2;
- `V*` collision for identical text with different images.

Controls passed:

- a tiny Qwen3.5 VLM processor generated `pixel_values`, `image_grid_thw`, and `mm_token_type_ids` on CPU;
- the existing A2PO text-only path completed Stage 1 and one Stage 2 optimizer step with finite loss.

The affected implementation remains unchanged on current main `3f738b90ed9de71c76f0593f575ad62851c30ba1`. I found no existing A2PO VLM issue or PR.

The important design question is the stable source-sample identity for Stage 1 caching and filtering. It must survive DataLoader sharding, distributed gather, filtering, and Stage 2 lookup. Rendered prompt text is insufficient for VLMs.

Possible directions:

1. an internal row ID assigned before Stage 1;
2. an explicit dataset ID column;
3. a trainer-internal fingerprint.

I prefer an internal row ID unless maintainers want the identity to be user-visible. Hashing image bytes does not seem like a good default because of its cost and remote/lazy image representations.

This proposal is separate from #6831, which discusses an explicit reference-model lifecycle for ZeRO-3. The first VLM implementation should not claim ZeRO-3 support.

Questions for maintainers:

1. Is the single-image, Transformers-generation, full-model scope appropriate for a first PR?
2. Is a trainer-internal source-row ID the preferred identity for `V*` caching, filtering, distributed gather, and Stage 2 lookup?
3. Should supported non-ZeRO-3 distributed execution be included in the first PR or follow after the single-process path?

### Your contribution

I can implement the agreed scope and add tests for Stage 1 image association across N samples, duplicate-text/different-image identity, filtering integrity, Stage 2 policy/reference multimodal forwards, completion-only finite loss, truncation safety, and text-only regression.

I have prepared executable CPU reproductions and positive controls. AI assistance was used during investigation and drafting; I reviewed the source analysis, reproductions, and proposed design directly.

**Comment by DimensionSTP** [comment]
@qgallouedec, could you advise on the sample-identity design before I proceed with an implementation?

Current A2PO still keys Stage 1 optimal values and filtering by rendered prompt text. For VLM support, identical text paired with different images must remain distinct throughout Stage 1, distributed collection, filtering, and Stage 2 lookup.

Would a trainer-internal source-row ID assigned before Stage 1 be the preferred approach? The initial scope would remain single-image, Transformers generation, and full-model training, without claiming ZeRO-3 support. This is separate from the reference-model API proposed in #6831. Thanks!
